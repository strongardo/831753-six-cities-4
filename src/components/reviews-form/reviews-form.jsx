import React from "react";
import {connect} from "react-redux";
import {setReviewsAsync} from "../../reducer/data/data.js";
import PropTypes from "prop-types";

const ReviewsForm = (props) => {
  const formRef = React.createRef();
  const buttonRef = React.createRef();

  let starsCount = 0;
  let text = ``;

  const toggleFormStatus = () => {
    formRef.current.querySelectorAll(`input, textarea, button`)
      .forEach((elem) => {
        elem.disabled = !elem.disabled;
      });
  };

  const onSuccess = () => {
    const form = formRef.current;
    toggleFormStatus(form);
    form.reset();
    starsCount = 0;
    buttonRef.current.disabled = true;
  };

  const onError = () => {
    const form = formRef.current;
    toggleFormStatus(form);
    form.querySelector(`textarea`).style.borderColor = `red`;
    setTimeout(() => {
      form.querySelector(`textarea`).style.borderColor = `#e6e6e6`;
    }, 1000);
  };

  const onSubmit = (evt) => {
    evt.preventDefault();
    toggleFormStatus();

    props.onSubmit(props.id, starsCount, text, onSuccess, onError);
  };

  const onRadioChange = (evt) => {
    starsCount = evt.target.defaultValue;

    if (text.length >= 50 && text.length < 300) {
      buttonRef.current.disabled = false;
    }
  };

  const onTextAriaChange = (evt) => {
    text = evt.target.value;

    if (text.length >= 50 && text.length < 300) {
      if (starsCount) {
        buttonRef.current.disabled = false;
      }
    } else {
      buttonRef.current.disabled = true;
    }
  };

  return (
    <form
      className="reviews__form form"
      action="#" method="post"
      onSubmit={onSubmit}
      ref = {formRef}
    >
      <label className="reviews__label form__label" htmlFor="review">
        Your review
      </label>
      <div className="reviews__rating-form form__rating">
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          defaultValue={5}
          id="5-stars"
          type="radio"
          onChange={onRadioChange}
        />
        <label
          htmlFor="5-stars"
          className="reviews__rating-label form__rating-label"
          title="perfect"
        >
          <svg
            className="form__star-image"
            width={37}
            height={33}
          >
            <use xlinkHref="#icon-star" />
          </svg>
        </label>
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          defaultValue={4}
          id="4-stars"
          type="radio"
          onChange={onRadioChange}
        />
        <label
          htmlFor="4-stars"
          className="reviews__rating-label form__rating-label"
          title="good"
        >
          <svg
            className="form__star-image"
            width={37}
            height={33}
          >
            <use xlinkHref="#icon-star" />
          </svg>
        </label>
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          defaultValue={3}
          id="3-stars"
          type="radio"
          onChange={onRadioChange}
        />
        <label
          htmlFor="3-stars"
          className="reviews__rating-label form__rating-label"
          title="not bad"
        >
          <svg
            className="form__star-image"
            width={37}
            height={33}
          >
            <use xlinkHref="#icon-star" />
          </svg>
        </label>
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          defaultValue={2}
          id="2-stars"
          type="radio"
          onChange={onRadioChange}
        />
        <label
          htmlFor="2-stars"
          className="reviews__rating-label form__rating-label"
          title="badly"
        >
          <svg
            className="form__star-image"
            width={37}
            height={33}
          >
            <use xlinkHref="#icon-star" />
          </svg>
        </label>
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          defaultValue={1}
          id="1-star"
          type="radio"
          onChange={onRadioChange}
        />
        <label
          htmlFor="1-star"
          className="reviews__rating-label form__rating-label"
          title="terribly"
        >
          <svg
            className="form__star-image"
            width={37}
            height={33}
          >
            <use xlinkHref="#icon-star" />
          </svg>
        </label>
      </div>
      <textarea
        className="reviews__textarea form__textarea"
        id="review"
        name="review"
        placeholder="Tell how was your stay, what you like and what can be improved"
        defaultValue={``}
        onInput={onTextAriaChange}
      />
      <div className="reviews__button-wrapper">
        <p className="reviews__help">
          To submit review please make sure to set{` `}
          <span className="reviews__star">rating</span> and
          describe your stay with at least{` `}
          <b className="reviews__text-amount">50 characters</b>.
        </p>
        <button
          className="reviews__submit form__submit button"
          type="submit"
          disabled={true}
          ref={buttonRef}
        >
          Submit
        </button>
      </div>
    </form>
  );
};

ReviewsForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onSubmit(id, starsCount, text, onSuccess, onError) {
    dispatch(setReviewsAsync(id, {
      comment: text,
      rating: starsCount,
    }, onSuccess, onError));
  },
});


export {ReviewsForm};
export default connect(null, mapDispatchToProps)(ReviewsForm);
