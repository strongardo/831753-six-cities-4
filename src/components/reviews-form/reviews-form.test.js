import React from "react";
import renderer from "react-test-renderer";
import {ReviewsForm} from "./reviews-form.jsx";

it(`Should ReviewsForm render correctly`, () => {
  const tree = renderer
    .create(
        <ReviewsForm
          text={`The house among olive`}
          starsCount={5}
          allDisabled={false}
          isError={false}
          onTextAriaChange={()=>{}}
          onRadioChange={()=>{}}
          allDisabledChange={()=>{}}
          onErrorForAlert={()=>{}}
          sendReview={()=>{}}
          id={1}
        />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
