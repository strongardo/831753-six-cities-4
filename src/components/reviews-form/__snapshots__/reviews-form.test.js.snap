// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewsForm render correctly 1`] = `
<form
  className="reviews__form form"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <p
    style={
      Object {
        "color": "red",
      }
    }
  />
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={true}
      className="form__rating-input visually-hidden"
      disabled={false}
      id="0-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="0-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled={false}
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled={false}
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled={false}
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled={false}
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled={false}
    id="review"
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value="The house among olive"
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set
       
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least
       
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
